test/test_main.py

```diff
@@ -0,0 +1,145 @@
+@patch.object(main, 'get_ftp_url')
+@patch.object(main, 'get_ftp_file')
+@patch.object(main, 'parse_xml')
+@patch.object(main, 'get_wzdx_schema')
+def test_translate_newest_icone_to_wzdx(get_wzdx_schema, parse_xml, get_ftp_file, get_ftp_url, pubsub):
+#the intent of this magic mock fuction is that we give a valid input ,that publishes data
+    main.get_ftp_url=MagicMock(return_value='')
+    main.get_ftp_file=MagicMock(return_value='')
+    main.parse_xml=MagicMock(return_value='')
+    main.get_wzdx_schema=MagicMock(return_value='')
+    icone_translator.wzdx_creator= MagicMock(return_value='WZDx')
+    icone_translator.validate_wzdx= MagicMock(return_value=True)
+    os.environ['project_id']='project_id'
+    os.environ['wzdx_topic_id']='wzdx_topic_id'
+    main.translate_newest_icone_to_wzdx(None,None)
+    publisher=pubsub().publish
+    publisher.assert_called_with(pubsub().topic_path('project_id', 'wzdx_topic_id'),str.encode(json.dumps('WZDx', indent=2)),origin='auto_icone_translator_ftp cloud function')
+
+
+@patch('google.cloud.pubsub_v1.PublisherClient')
+@patch.object(main, 'get_ftp_url')
+@patch.object(main, 'get_ftp_file')
+@patch.object(main, 'parse_xml')
+@patch.object(main, 'get_wzdx_schema')
+def test_translate_newest_icone_to_wzdx_with_invalid_data(get_wzdx_schema, parse_xml, get_ftp_file, get_ftp_url, pubsub):
+#the intent of this magic mock fuction is that we give a valid input ,that publishes data
+    main.get_ftp_url=MagicMock(return_value='')
+    main.get_ftp_file=MagicMock(return_value='')
+    main.parse_xml=MagicMock(return_value='')
+    main.get_wzdx_schema=MagicMock(return_value='')
+    icone_translator.wzdx_creator= MagicMock(return_value='WZDx')
+    icone_translator.validate_wzdx= MagicMock(return_value=False)
+    os.environ['project_id']='project_id'
+    os.environ['wzdx_topic_id']='wzdx_topic_id'
+    output=main.translate_newest_icone_to_wzdx(None,None)
+    publisher=pubsub().publish
+    publisher.assert_not_called()
+
+
+
+@patch.object(main, 'get_ftp_credentials')
+def test_get_ftp_url(ftp_credentials):
+    credentials='username', 'password'
+    main.get_ftp_credentials=MagicMock(return_value=credentials)
+    os.environ['ftp_server_address']='www.icone.com'
+    os.environ['ftp_port']='4425'
+    os.environ['icone_ftp_username']='username'
+    os.environ['icone_ftp_password']='password'
+    os.environ['ftp_icone_file_path']='test_filepath'
+    test_ftp_url='ftp://username:password@www.icone.com:4425/test_filepath'
+    actual=main.get_ftp_url()
+    del os.environ['ftp_server_address']
+    del os.environ['ftp_port']
+    del os.environ['icone_ftp_username']
+    del os.environ['icone_ftp_password']
+    del os.environ['ftp_icone_file_path']
+    assert actual==test_ftp_url
+
+@patch.object(main, 'get_ftp_credentials')
+def test_get_ftp_url_invalid_data(ftp_credentials):
+    credentials='username', 'password'
+    main.get_ftp_credentials=MagicMock(return_value=credentials)
+    try:
```

Comment:
As in `test_translator.py`, we should be using the unit testing tools to verify this exception is thrown rather than surrounding in try/catch

Reply:
the function is updated to not throw an exception.

Reply:
i used the unit test tools and verified the function  and updated the unit test